File: mock/front_sim.py
"""Synthetic front obstacle fixture, never physical calibration. Matches browser demo."""
import math

class FrontSim:
    def __init__(self):
        self.distance=120
        self.phase='NONE'
        self.enabled=self.held=self.valid=self.blocked=False
        self.reason='boot'
        self.at=self.right_at=0
        self.sample_at=-1
        self.near=self.clear=0
        self.samples=[]
        self.raw=self.filtered=120

    def cancel(self, reason):
        self.phase='NONE'
        self.enabled=False
        self.reason=reason

    def set_demo(self, enabled, t):
        self.cancel('demo_changed')
        self.enabled=enabled
        self.at=t

    def enter(self, phase, t):
        self.phase=phase
        self.at=t
        if phase=='RIGHT': self.right_at=t

    def step(self, t, target, following):
        zero=[0.,0.,0.]
        distance=self.distance
        if self.enabled and distance==120:
            distance=18 if self.phase in ('NONE','HALT') or self.phase=='RIGHT' and t-self.right_at<.6 else 120
        if t-self.sample_at>=.07:
            self.sample_at=t
            self.valid=type(distance) in (int,float) and math.isfinite(distance) and 2<=distance<=400
            self.raw=distance
            if self.valid:
                self.samples=(self.samples+[distance])[-3:]
                self.filtered=sorted(self.samples)[len(self.samples)//2]
                self.near=self.near+1 if distance<=20 else 0
                self.clear=self.clear+1 if distance>=70 else 0
                if distance<=20: self.blocked=True
                elif self.clear>=3: self.blocked=False
            else:
                self.samples=[]
                self.near=self.clear=0
        if not self.valid:
            self.cancel('front_unknown')
            return zero,True
        if following and self.phase=='NONE' and (target[0]>0 or self.enabled) and self.blocked:
            if not self.enabled or self.raw>20:
                self.cancel('front_obstacle')
                return zero,True
            self.enter('HALT',t)
        if self.phase=='HALT' and self.near>=3 and t-self.at>=.21: self.enter('RIGHT',t)
        elif self.phase=='RIGHT' and self.clear>=3: self.enter('MARGIN',t)
        elif self.phase=='MARGIN':
            if self.raw<70:
                self.phase='RIGHT'
                self.clear=0
            elif t-self.at>=.4: self.enter('PASS',t)
        elif self.phase=='PASS':
            if self.raw<50:
                self.cancel('bypass_blocked')
                return zero,True
            if t-self.at>=.8: self.enter('REACQUIRE',t)
        elif self.phase=='REACQUIRE' and t-self.at>=.3: self.cancel('bypass_complete')
        if self.phase in ('RIGHT','MARGIN') and t-self.right_at>=4:
            self.cancel('bypass_timeout')
            return zero,True
        if self.phase in ('HALT','REACQUIRE') and t-self.at>=3:
            self.cancel('front_unstable')
            return zero,True
        if self.phase in ('HALT','REACQUIRE'): return zero,False
        if self.phase in ('RIGHT','MARGIN'): return [0.,.15,0.],False
        if self.phase=='PASS': return [.15,0.,0.],False
        if self.held: return zero,False
        if target[0]>0 and self.blocked:
            self.held=True
            self.reason='front_obstacle'
            return zero,False
        target=target[:]
        if target[0]>0: target[0]*=max(0,min(1,(min(self.raw,self.filtered)-20)/30))
        return target,False

    def telemetry(self,t):
        status='UNKNOWN' if not self.valid else 'STOPPED' if self.held else 'BYPASS' if self.phase!='NONE' else 'BLOCKED' if self.blocked else 'SLOW' if self.filtered<50 else 'WARN' if self.filtered<60 else 'CLEAR'
        return dict(enabled=True,ready=True,valid=self.valid,distance_cm=self.filtered if self.valid else None,
            age_ms=max(0,(t-self.sample_at)*1000),demo_ready=True,demo_enabled=self.enabled,
            release_required=self.held,phase=self.phase,stop_reason=self.reason,status=status)

File: js/front-sim.js
// DEMONSTRATION FIXTURE ONLY. These numbers are never uploaded as physical calibration.
export class FrontSim {
  constructor() { this.distance=120;this.phase='NONE';this.enabled=false;this.held=false;this.blocked=false;this.reason='boot';this.at=0;this.rightAt=0;this.near=0;this.clear=0;this.samples=[];this.sampleAt=-1;this.valid=false;this.raw=120;this.filtered=120; }
  cancel(reason) { this.phase='NONE';this.enabled=false;this.reason=reason; }
  setDemo(enabled,t) { this.cancel('demo_changed');this.enabled=enabled;this.at=t; }
  release() { this.held=false; }
  enter(phase,t) { this.phase=phase;this.at=t;if(phase==='RIGHT')this.rightAt=t; }
  step(t,target,following) {
    const zero={vx:0,vy:0,wz:0};let abort=false;
    // Explicit distance override for faults; null means no echo, not open space.
    let distance=this.distance;
    if(this.enabled&&this.distance===120)distance=['NONE','HALT'].includes(this.phase)||this.phase==='RIGHT'&&t-this.rightAt<.6?18:120;
    if(t-this.sampleAt>=.07) {
      this.sampleAt=t;this.valid=Number.isFinite(distance)&&distance>=2&&distance<=400;this.raw=distance;
      if(this.valid) { this.samples.push(distance);if(this.samples.length>3)this.samples.shift();this.filtered=[...this.samples].sort((a,b)=>a-b)[Math.floor(this.samples.length/2)];this.near=distance<=20?this.near+1:0;this.clear=distance>=70?this.clear+1:0;if(distance<=20)this.blocked=true;else if(this.clear>=3)this.blocked=false; }
      else {this.samples=[];this.near=this.clear=0;}
    }
    if(!this.valid) { this.cancel('front_unknown');return {target:zero,abort:true}; }
    if(following&&this.phase==='NONE'&&(target.vx>0||this.enabled)&&this.blocked) {
      if(!this.enabled||this.raw>20) {this.cancel('front_obstacle');return {target:zero,abort:true};}
      this.enter('HALT',t);
    }
    if(this.phase==='HALT'&&this.near>=3&&t-this.at>=.21)this.enter('RIGHT',t);
    else if(this.phase==='RIGHT'&&this.clear>=3)this.enter('MARGIN',t);
    else if(this.phase==='MARGIN') {
      if(this.raw<70) {this.phase='RIGHT';this.clear=0;}
      else if(t-this.at>=.4)this.enter('PASS',t);
    } else if(this.phase==='PASS') {
      if(this.raw<50) {this.cancel('bypass_blocked');return {target:zero,abort:true};}
      if(t-this.at>=.8)this.enter('REACQUIRE',t);
    } else if(this.phase==='REACQUIRE'&&t-this.at>=.3)this.cancel('bypass_complete');
    if(['RIGHT','MARGIN'].includes(this.phase)&&t-this.rightAt>=4) {this.cancel('bypass_timeout');abort=true;}
    if(['HALT','REACQUIRE'].includes(this.phase)&&t-this.at>=3) {this.cancel('front_unstable');abort=true;}
    if(abort)return {target:zero,abort};
    if(['HALT','REACQUIRE'].includes(this.phase))return {target:zero};
    if(['RIGHT','MARGIN'].includes(this.phase))return {target:{vx:0,vy:.15,wz:0}};
    if(this.phase==='PASS')return {target:{vx:.15,vy:0,wz:0}};
    if(this.held)return {target:zero};
    if(target.vx>0&&this.blocked) {this.held=true;this.reason='front_obstacle';return {target:zero};}
    return {target:{...target,vx:target.vx>0?target.vx*Math.max(0,Math.min(1,(Math.min(this.raw,this.filtered)-20)/30)):target.vx}};
  }
  telemetry(t) {
    return { enabled:true,ready:true,valid:this.valid,distance_cm:this.valid?this.filtered:null,age_ms:Math.max(0,(t-this.sampleAt)*1000),
      demo_ready:true,demo_enabled:this.enabled,release_required:this.held,phase:this.phase,stop_reason:this.reason,
      status:!this.valid?'UNKNOWN':this.held?'STOPPED':this.phase!=='NONE'?'BYPASS':this.blocked?'BLOCKED':this.filtered<50?'SLOW':this.filtered<60?'WARN':'CLEAR' };
  }
}
